### PR TITLE
refactor: Rewrite `Clock` / add tests

### PR DESCRIPTION
**Pull request by pmnoxx**
As @matklad suggested, we will make `reset` and `mock` methods private. I also removed dead code.

Done:
- removed `add_interval` - this function was unused , and it was taking `Duration` as argument.
- removed unused `instant_call_count`
- unused `instant: Instant,`
- renamed `is_mock` to `is_mocked`
- made `reset()`, `mock()` private
- replaced `new` with `#[derive(Default)]` to make code simpler
- added comments
- removed unused `self.durations`
- added tests
- removed ` fn with<F, T>(f: F) -> T` it was a method which had `  static INSTANCE: `  hidden inside. Replaced with direct calls to `INSIDE`. This made it easier to document the code

Blocks https://github.com/near/nearcore/pull/5177

### CODE REVIEW

**Comment by pmnoxx** [comment]
@yordanmadzhunkov Can you take a look?

**Review by matklad — COMMENTED** [review]
LGTM!

Couple of more suggestions:

* we can replace `is_mock: bool,` state flag with storing `Option<MockClockPerThread>`. That way, we statically guarantee that we only have mocking-related state when we do in fact mock the clock
* move `add_utc` and `utc_call_count` methods from `Clock` to `MockClockGuard`. This again statically guarantees that we only can call these methods when mocking is in fact active.

File: core/primitives/src/time.rs [hunk 1]
```diff
@@ -9,62 +9,22 @@ use chrono::DateTime;
...
-            utc: VecDeque::with_capacity(16),
-            durations: VecDeque::with_capacity(16),
-            utc_call_count: 0,
-            instant_call_count: 0,
-            instant: Instant::now(),
-            is_mock: false,
-        }
-    }
+thread_local! {
+    static INSTANCE: RefCell<MockClockPerThread> = RefCell::default()
```
[the remark this answers is not in the document]
Comment (pmnoxx) [reply core/primitives/src/time.rs:27]: I see your point. In that case let's use `Type::with`. I agree that `thread_local` is a bit weird.

**Comment by pmnoxx** [comment]
@matklad @mm-near Can you take a look again?

**Review by matklad — APPROVED** [review]

File: core/primitives/src/time.rs [hunk 2]
```diff
@@ -82,74 +74,107 @@ impl Drop for MockClockGuard {
...
-            if clock.is_mock {
-                clock.durations.push_back(mock_instant);
-            } else {
-                panic!("Use MockClockGuard in your test");
-            }
-        });
+    /// Gets mocked instant.
+    pub fn instant() -> Instant {
+        MockClockPerThread::with(|clock| match &mut clock.mock {
+            Some(_clock) => panic!("Mock clock run out of samples"),
```
Comment (mm-near) [inline core/primitives/src/time.rs:93]: I'm confused - shouldn't this say 'not implemented'?

File: core/primitives/src/time.rs [hunk 3]
```diff
@@ -82,74 +74,107 @@ impl Drop for MockClockGuard {
...
+            None => Instant::now(),
+        })
     }
 
+    /// Returns time pushed by `Self::push_utc()`
```
[the remark this answers is not in the document]
Comment (mm-near) [reply core/primitives/src/time.rs:98]: and actually we should update this comment:

as this method returns current utc (if mocking is disabled) or time pushed by 'add_utc' if mocking is enabled

File: core/primitives/src/time.rs [hunk 4]
```diff
@@ -82,74 +74,107 @@ impl Drop for MockClockGuard {
     }
 }
 
+pub struct Clock {}
```
Comment (mm-near) [inline core/primitives/src/time.rs:77]: please comment that the 'Clock' is the class that you're supposed to pass to your 'production' code - and use its methods to fetch time.

File: core/primitives/src/time.rs [hunk 5]
```diff
@@ -82,74 +74,107 @@ impl Drop for MockClockGuard {
...
+            );
+            assert_eq!(
+                Clock::utc(),
+                utc_now
+                    .checked_add_signed(chrono::Duration::from_std(Duration::from_secs(3)).unwrap())
+                    .unwrap(),
+            );
+
+            assert_eq!(mock_clock_guard.utc_call_count(), 3);
+            drop(mock_clock_guard);
```
Comment (mm-near) [inline core/primitives/src/time.rs:175]: do you have to call it explicitly?

File: core/primitives/src/time.rs [hunk 6]
```diff
@@ -82,74 +74,107 @@ impl Drop for MockClockGuard {
...
+        }
+        let mock_clock_guard = MockClockGuard::default();
+        assert_eq!(mock_clock_guard.utc_call_count(), 0);
     }
```
Comment (mm-near) [inline core/primitives/src/time.rs:179]: can you also add a simple test that checks threading behavior?

**Review by mm-near — DISMISSED** [review]

File: core/primitives/src/time.rs [hunk 7]
```diff
@@ -9,27 +9,21 @@ use chrono::DateTime;
...
 use std::collections::VecDeque;
 
-struct MockClockPerThread {
+#[derive(Default)]
+struct MockClockPerState {
+    /// List of timestamp, we will return one timestamp to each call.
     utc: VecDeque<DateTime<Utc>>,
-    durations: VecDeque<Duration>,
+    /// Number of times `utc()` method was called since we started mocking.
     utc_call_count: u64,
```
[the remark this answers is not in the document]
Comment (pmnoxx) [reply core/primitives/src/time.rs:17]: Ok, I brought it back.

File: core/primitives/src/time.rs (hunk 5 above)
Comment (pmnoxx) [reply core/primitives/src/time.rs:175]: No, this was done, so we can create another mock_clock_guard. You can have only one active at a time.

File: core/primitives/src/time.rs (hunk 6 above)
Comment (pmnoxx) [reply core/primitives/src/time.rs:226]: I added a test.

File: core/primitives/src/time.rs [hunk 8]
```diff
@@ -39,36 +33,34 @@ impl MockClockPerThread {
...
-    }
-    fn pop_instant(&mut self) -> Option<Instant> {
-        self.instant_call_count += 1;
-        let x = self.durations.pop_front();
-        match x {
-            Some(t) => self.instant.checked_add(t),
-            None => None,
-        }
+pub struct MockClockGuard {}
+
```
[the remark this answers is not in the document]
Comment (pmnoxx) [reply core/primitives/src/time.rs:43]: Ok, I added usage example.

File: core/primitives/src/time.rs [hunk 9]
```diff
@@ -1,35 +1,51 @@
...
 use std::collections::VecDeque;
+use std::default::Default;
+pub use std::time::{Duration, Instant};
 
-struct MockClockPerThread {
-    utc: VecDeque<DateTime<Utc>>,
-    durations: VecDeque<Duration>,
+#[derive(Default)]
+struct MockClockPerState {
+    /// List of timestamps, we will return one timestamp to each call.
```
Comment (mm-near) [inline core/primitives/src/time.rs:32]: /to each call/ for each call to Clock::utc()

File: core/primitives/src/time.rs [hunk 10]
```diff
@@ -1,35 +1,51 @@
...
+    utc_list: VecDeque<DateTime<Utc>>,
+    /// List of timestamps, we will return one timestamp to each call.
```
Comment (mm-near) [inline core/primitives/src/time.rs:34]: for Clock::now

File: core/primitives/src/time.rs [hunk 11]
```diff
@@ -1,35 +1,51 @@
-use std::default::Default;
-
+/// Provides structs used for getting time.
+///
```
Comment (mm-near) [inline core/primitives/src/time.rs:2]: I'd also mention that people should be using Clock whenever possible - and write some guideline on when they should use now() vs instant()

**Review by mm-near — APPROVED** [review]

File: core/primitives/src/time.rs (hunk 11 above)
Comment (pmnoxx) [reply core/primitives/src/time.rs:2]: @mm-near My personal opinion on this is that we should add a large disclaimer, that the usage of `Clock` is incomplete.
And, it should be discouraged. 

```
   let i = Clock::instant();
   /// This get's current time, not mocked time
   i.elapsed().
```

Comment (pmnoxx) [reply core/primitives/src/time.rs:2]:
```
/// WARNING WARNING WARNING
/// WARNING WARNING WARNING
/// Use at your own risk. The implementation is not complete, we have a places in code not mocked properly.
/// For example, it's possible to call `::elapsed()` on `Instant` returned by `Instant::now()`.
/// We use that that function, throughout the code, and the current mocking of `Instant` is not done properly.
```

**Review by yordanmadzhunkov — APPROVED** [review]
Awesome refactoring.

File: core/primitives/src/time.rs [hunk 12]
```diff
@@ -82,74 +123,165 @@ impl Drop for MockClockGuard {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::ops::Add;
+    use std::thread;
+    use std::thread::sleep;
+
+    #[test]
+    #[should_panic]
```
Comment (yordanmadzhunkov) [inline core/primitives/src/time.rs:189]: I just want to mantion, that this test is awesome :) Good job

**Comment by pmnoxx** [comment]
> Awesome refactoring.

Thanks, for introducing the time mocking into our repo. That helps us a lot!

**Pull request merged**